File: flutter_app/lib/services/friends_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'api_service.dart';

class FriendUser {
  final String uid;
  final String username;
  final String displayName;
  final String? photoUrl;
  final int gamesPlayed;
  final int gamesWon;
  final int averageScore;
  final bool isOnline;

  const FriendUser({
    required this.uid,
    required this.username,
    required this.displayName,
    this.photoUrl,
    this.gamesPlayed = 0,
    this.gamesWon = 0,
    this.averageScore = 0,
    this.isOnline = false,
  });

  factory FriendUser.fromJson(Map<String, dynamic> json) {
    return FriendUser(
      uid: json['uid'] as String? ?? '',
      username: json['username'] as String? ?? json['displayName'] as String? ?? 'user',
      displayName: json['displayName'] as String? ?? 'Player',
      photoUrl: json['photoUrl'] as String?,
      gamesPlayed: json['gamesPlayed'] as int? ?? 0,
      gamesWon: json['gamesWon'] as int? ?? 0,
      averageScore: json['averageScore'] as int? ?? 0,
      isOnline: json['isOnline'] as bool? ?? false,
    );
  }
}

class FriendRequest {
  final String id;
  final String fromUid;
  final String fromDisplayName;
  final String fromUsername;
  final String createdAt;

  const FriendRequest({
    required this.id,
    required this.fromUid,
    required this.fromDisplayName,
    required this.fromUsername,
    required this.createdAt,
  });

  factory FriendRequest.fromJson(Map<String, dynamic> json) {
    return FriendRequest(
      id: json['id'] as String? ?? '',
      fromUid: json['fromUid'] as String? ?? '',
      fromDisplayName: json['fromDisplayName'] as String? ?? 'Player',
      fromUsername: json['fromUsername'] as String? ?? 'user',
      createdAt: json['createdAt'] as String? ?? '',
    );
  }
}

class FriendsService {
  final String baseUrl;
  final String Function() getToken;

  FriendsService({required this.baseUrl, required this.getToken});

  Map<String, String> get _headers => {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer ${getToken()}',
      };

  Future<List<FriendUser>> searchUsers(String query) async {
    final uri = Uri.parse('$baseUrl/api/friends/search?query=${Uri.encodeComponent(query)}');
    final response = await http.get(uri, headers: _headers);
    final data = jsonDecode(response.body);
    if (!data['success']) throw Exception(data['error']?['message'] ?? 'Failed to search users');
    final list = data['data']['users'] as List;
    return list.map((u) => FriendUser.fromJson(u)).toList();
  }

  Future<List<FriendUser>> getFriends() async {
    final uri = Uri.parse('$baseUrl/api/friends');
    final response = await http.get(uri, headers: _headers);
    final data = jsonDecode(response.body);
    if (!data['success']) throw Exception(data['error']?['message'] ?? 'Failed to get friends');
    final list = data['data']['friends'] as List;
    return list.map((u) => FriendUser.fromJson(u)).toList();
  }

  Future<Map<String, List<FriendRequest>>> getPendingRequests() async {
    final uri = Uri.parse('$baseUrl/api/friends/requests');
    final response = await http.get(uri, headers: _headers);
    final data = jsonDecode(response.body);
    if (!data['success']) throw Exception(data['error']?['message'] ?? 'Failed to get requests');

    final incomingList = data['data']['incoming'] as List? ?? [];
    final incoming = incomingList.map((r) => FriendRequest.fromJson(r)).toList();

    return {'incoming': incoming};
  }

  Future<void> sendRequest(String targetUsernameOrUid) async {
    final uri = Uri.parse('$baseUrl/api/friends/request');
    final response = await http.post(
      uri,
      headers: _headers,
      body: jsonEncode({'target': targetUsernameOrUid}),
    );
    final data = jsonDecode(response.body);
    if (response.statusCode != 200 || !data['success']) {
      throw Exception(data['error']?['message'] ?? 'Failed to send request');
    }
  }

  Future<void> acceptRequest(String requesterId) async {
    final uri = Uri.parse('$baseUrl/api/friends/accept');
    final response = await http.post(
      uri,
      headers: _headers,
      body: jsonEncode({'requesterId': requesterId}),
    );
    final data = jsonDecode(response.body);
    if (!data['success']) throw Exception(data['error']?['message'] ?? 'Failed to accept request');
  }

  Future<void> declineRequest(String targetUid) async {
    final uri = Uri.parse('$baseUrl/api/friends/decline');
    final response = await http.post(
      uri,
      headers: _headers,
      body: jsonEncode({'targetUid': targetUid}),
    );
    final data = jsonDecode(response.body);
    if (!data['success']) throw Exception(data['error']?['message'] ?? 'Failed to decline request');
  }

  Future<void> removeFriend(String friendId) async {
    final uri = Uri.parse('$baseUrl/api/friends/$friendId');
    final response = await http.delete(uri, headers: _headers);
    final data = jsonDecode(response.body);
    if (!data['success']) throw Exception(data['error']?['message'] ?? 'Failed to remove friend');
  }
}

File: backend/src/services/friends.service.js
import { getFirestore } from '../config/firebase.js';
import logger from '../utils/logger.js';

/**
 * Friends Service — handles user search, friend requests, and friend list management.
 */
class FriendsService {
  /**
   * Search users by username or display name (case insensitive).
   */
  async searchUsers(query, currentUserId) {
    if (!query || query.trim().length === 0) return [];

    const db = getFirestore();
    const cleanQuery = query.trim().toLowerCase();

    const snapshot = await db.collection('users').get();
    const users = [];

    snapshot.docs.forEach((doc) => {
      const u = doc.data();
      if (u.uid === currentUserId) return;

      const username = (u.username || '').toLowerCase();
      const displayName = (u.displayName || '').toLowerCase();

      if (username.includes(cleanQuery) || displayName.includes(cleanQuery)) {
        users.push({
          uid: u.uid,
          username: u.username || u.displayName,
          displayName: u.displayName || 'Player',
          photoUrl: u.photoUrl || null,
          gamesPlayed: u.gamesPlayed || 0,
          gamesWon: u.gamesWon || 0,
          averageScore: u.averageScore || 0,
        });
      }
    });

    return users.slice(0, 20);
  }

  /**
   * Send a friend request to a target user (by UID or Username).
   */
  async sendFriendRequest(fromUserId, targetQuery) {
    const db = getFirestore();
    const fromDoc = await db.collection('users').doc(fromUserId).get();
    if (!fromDoc.exists) throw new Error('User not found');
    const fromUser = fromDoc.data();

    // Find target user
    let targetUser = null;
    const allUsersSnap = await db.collection('users').get();
    const targetClean = targetQuery.trim().toLowerCase();

    for (const doc of allUsersSnap.docs) {
      const u = doc.data();
      if (u.uid === targetQuery || (u.username && u.username.toLowerCase() === targetClean) || (u.displayName && u.displayName.toLowerCase() === targetClean)) {
        targetUser = u;
        break;
      }
    }

    if (!targetUser) {
      throw new Error(`Player "${targetQuery}" not found.`);
    }

    if (targetUser.uid === fromUserId) {
      throw new Error('You cannot send a friend request to yourself.');
    }

    // Check if already friends
    const fromFriends = fromUser.friends || [];
    if (fromFriends.includes(targetUser.uid)) {
      throw new Error(`You are already friends with ${targetUser.displayName}.`);
    }

    // Check if request already exists
    const requestId = `${fromUserId}_${targetUser.uid}`;
    const reverseRequestId = `${targetUser.uid}_${fromUserId}`;

    const existingReq = await db.collection('friend_requests').doc(requestId).get();
    if (existingReq.exists) {
      throw new Error('Friend request already sent.');
    }

    const reverseReq = await db.collection('friend_requests').doc(reverseRequestId).get();
    if (reverseReq.exists) {
      // Auto-accept if they already sent us a request!
      return this.acceptFriendRequest(fromUserId, targetUser.uid);
    }

    const newRequest = {
      id: requestId,
      fromUid: fromUserId,
      fromDisplayName: fromUser.displayName || 'Player',
      fromUsername: fromUser.username || fromUser.displayName,
      toUid: targetUser.uid,
      toDisplayName: targetUser.displayName || 'Player',
      toUsername: targetUser.username || targetUser.displayName,
      status: 'pending',
      createdAt: new Date().toISOString(),
    };

    await db.collection('friend_requests').doc(requestId).set(newRequest);
    logger.info(`Friend request sent: ${fromUser.displayName} -> ${targetUser.displayName}`);

    return newRequest;
  }

  /**
   * Accept a friend request.
   */
  async acceptFriendRequest(currentUserId, requesterId) {
    const db = getFirestore();
    const requestId = `${requesterId}_${currentUserId}`;

    const reqDoc = await db.collection('friend_requests').doc(requestId).get();
    if (reqDoc.exists) {
      await db.collection('friend_requests').doc(requestId).set({ status: 'accepted' });
    }

    // Update user 1 friends list
    const user1Ref = db.collection('users').doc(currentUserId);
    const user1Doc = await user1Ref.get();
    if (user1Doc.exists) {
      const u1Data = user1Doc.data();
      const friends = new Set(u1Data.friends || []);
      friends.add(requesterId);
      await user1Ref.update({ friends: Array.from(friends) });
    }

    // Update user 2 friends list
    const user2Ref = db.collection('users').doc(requesterId);
    const user2Doc = await user2Ref.get();
    if (user2Doc.exists) {
      const u2Data = user2Doc.data();
      const friends = new Set(u2Data.friends || []);
      friends.add(currentUserId);
      await user2Ref.update({ friends: Array.from(friends) });
    }

    logger.info(`Friend request accepted between ${currentUserId} and ${requesterId}`);
    return { success: true };
  }

  /**
   * Decline or cancel a friend request.
   */
  async cancelOrDeclineRequest(fromUid, toUid) {
    const db = getFirestore();
    const req1 = `${fromUid}_${toUid}`;
    const req2 = `${toUid}_${fromUid}`;

    await db.collection('friend_requests').doc(req1).set({ status: 'declined' });
    await db.collection('friend_requests').doc(req2).set({ status: 'declined' });

    return { success: true };
  }

  /**
   * Remove a friend.
   */
  async removeFriend(currentUserId, friendId) {
    const db = getFirestore();

    const u1Ref = db.collection('users').doc(currentUserId);
    const u1Doc = await u1Ref.get();
    if (u1Doc.exists) {
      const friends = (u1Doc.data().friends || []).filter((id) => id !== friendId);
      await u1Ref.update({ friends });
    }

    const u2Ref = db.collection('users').doc(friendId);
    const u2Doc = await u2Ref.get();
    if (u2Doc.exists) {
      const friends = (u2Doc.data().friends || []).filter((id) => id !== currentUserId);
      await u2Ref.update({ friends });
    }

    return { success: true };
  }

  /**
   * Get current user's friends details.
   */
  async getFriends(userId) {
    const db = getFirestore();
    const userDoc = await db.collection('users').doc(userId).get();
    if (!userDoc.exists) return [];

    const friendIds = userDoc.data().friends || [];
    if (friendIds.length === 0) return [];

    const friends = [];
    for (const fId of friendIds) {
      const fDoc = await db.collection('users').doc(fId).get();
      if (fDoc.exists) {
        const u = fDoc.data();
        friends.push({
          uid: u.uid,
          username: u.username || u.displayName,
          displayName: u.displayName || 'Player',
          photoUrl: u.photoUrl || null,
          gamesPlayed: u.gamesPlayed || 0,
          gamesWon: u.gamesWon || 0,
          averageScore: u.averageScore || 0,
          isOnline: u.isOnline || false,
        });
      }
    }

    return friends;
  }

  /**
   * Get pending friend requests (incoming & outgoing).
   */
  async getPendingRequests(userId) {
    const db = getFirestore();
    const reqSnap = await db.collection('friend_requests').get();

    const incoming = [];
    const outgoing = [];

    reqSnap.docs.forEach((doc) => {
      const r = doc.data();
      if (r.status !== 'pending') return;

      if (r.toUid === userId) {
        incoming.push({
          id: r.id,
          fromUid: r.fromUid,
          fromDisplayName: r.fromDisplayName,
          fromUsername: r.fromUsername,
          createdAt: r.createdAt,
        });
      } else if (r.fromUid === userId) {
        outgoing.push({
          id: r.id,
          toUid: r.toUid,
          toDisplayName: r.toDisplayName,
          toUsername: r.toUsername,
          createdAt: r.createdAt,
        });
      }
    });

    return { incoming, outgoing };
  }
}

export default new FriendsService();
